### PR TITLE
reproduce bug regarding duplicate type definitions from local references

### PR DESCRIPTION
**Pull request by chriskn**
See Issue #130

### CODE REVIEW

**Comment by ImFlog** [comment]
I spent some time on this tonight and I have questions for you.
Following this [stackoverflow post](https://stackoverflow.com/questions/48100575/avro-multiple-record-of-same-type-in-single-schema) I think we can reference the schema with the name = "${namespace}.${typeRef}".

Based on the test case you provided, it would give something like this :

```json
{
      "name": "A",
      "namespace": "com.mycompany",
      "type": "record",
      "fields":
        [
          {
            "name": "nested",
            "type":
              {
                "name": "B",
                "namespace": "com.mycompany",
                "type": "enum",
                "symbols": ["X1", "X2"]
              }
          },
          {
            "name": "nested1",
            "type":
              {
                "name": "com.mycompany.B",
                "type": "enum"
              }
          }
        ]
    }
```

WDYT ?

I still have some challenge to tackle around the namespace retrieval (retrieve the namespace of the localReference and if not available use the global namespace). Plus the fact that it's a recursive code that is not that easy to understand 😅 


Also CC @dejank1986 as it would solve #126 too.

**Comment by chriskn** [comment]
The version 1.8.0 would not use the namespace for referencing type B. Because the B has the same Namespace then A. The schema generated with 1.8.0 would look like this:

```
{
  "name": "A",
  "namespace": "com.mycompany",
  "type": "record",
  "fields": [
    {
      "name": "nested",
      "type": {
        "name": "B",
        "type": "enum",
        "symbols": ["X1", "X2"]
      }
    },
    {
      "name": "nested1",
      "type": "B"
    }
  ]
}
```

Is it possible to compare the namespace of a reference with the namespace of the subject and if they are the same, skip the namespace when referencing?

**Comment by ImFlog** [comment]
Thanks for the input, I understand better what It should look like.

If the namespace is not specified we can indeed not specify it, but in case it's defined in the above local reference I guess I need to specify It in the `type` field.

I will update the test you provided to reflect that, I still need to add a bit of manual parsing of the content to verify if the namespace is available or not but it's clear in my head (I think).

Will have a bit of time early next week I think.

**Comment by ImFlog** [comment]
Hello @chriskn, the tests are green and I managed the cases around namespace management as requested.
I am honestly not very fan of the code, it's a bit hard to follow (like most of the recursive ones I guess ...), but at least it works.
If you want to give it a spin locally or do a review that would be great, I want to merge this soon as you are not the only one impacted.
Thanks

**Comment by chriskn** [comment]
Hello @ImFlog, thank you for the fix! I will test it as soon as possible, latest tomorrow.

**Comment by chriskn** [comment]
I am currently testing the new version and getting the following exception: 

```
Caused by: org.json.JSONException: JSONObject["name"] not found.
	at org.json.JSONObject.get(JSONObject.java:587)
	at org.json.JSONObject.getString(JSONObject.java:874)
	at com.github.imflog.schema.registry.parser.AvroSchemaParser.replaceType(AvroSchemaParser.kt:64)
	at com.github.imflog.schema.registry.parser.AvroSchemaParser.replaceType(AvroSchemaParser.kt:71)
	at com.github.imflog.schema.registry.parser.AvroSchemaParser.replaceType(AvroSchemaParser.kt:58)
	at com.github.imflog.schema.registry.parser.AvroSchemaParser.replaceType(AvroSchemaParser.kt:71)
	at com.github.imflog.schema.registry.parser.AvroSchemaParser.resolveLocalReferences(AvroSchemaParser.kt:42)
	at com.github.imflog.schema.registry.parser.SchemaParser.parseSchemaFromFile(SchemaParser.kt:39)
	at com.github.imflog.schema.registry.tasks.compatibility.CompatibilityTaskAction.run(CompatibilityTaskAction.kt:29)
	at com.github.imflog.schema.registry.tasks.compatibility.CompatibilityTask.testCompatibility(CompatibilityTask.kt:44)
```

I will look into it

**Comment by chriskn** [comment]
@ImFlog, I ran into multiple issues with the current implementation 

Array type references are not supported
Types with additional, java specific information like  
```
     "items": {
        "type": "string",
        "avro.java.string": "String"
      }
```
are not supported. 

I will try to fix these issues

**Comment by chriskn** [comment]
I invested some hours today and came to the conclusion that it does not make sense to reimplement the schema parsing. Instead, the `Schema.Parser` provided by `org.apache.avro` should be used. 

In the current state, this plugin is completely broken regarding schema parsing.
Therefore, I reverted the `AvroSchemaParser` to version 1.8.0.

One test is failing afterward. The reason is that the `Street` record is referenced by a local schema, but not included when parsing the schema. This issue did not occur in the latest version because of other issues with the schema parsing.

Since I don't see much benefit from referencing remote schemas (then can be downloaded before using the download task) I did not fix this issue. 

 It's up to you how to go further from here.

**Comment by ImFlog** [comment]
@chriskn we are not reimplementing the Avro parsing, only replacing local references before doing the normal parsing. The goal is to be able to handle local + remote dependencies for Avro schemas.
There are some edge-cases but I will try to fix them (I will revert your last change back).

In the meantime you can stick to 1.8.0 and hopefully I will be able to fix this during the weekend.

Thank you for spending time on this PR 🙇‍♂️

**Comment by ImFlog** [comment]
@chriskn Can you explain a bit the issue with the custom java types ?

**Comment by ImFlog** [comment]
After thinking a bit about what you say and pocking things, I think that the easier is to fork the avro parser in order to silently ignore the dependencies it doesn't know.
I've done a draft locally and it seems to work for all the use cases. As the project is not moving very fast I should be able to keep up with the updates easily.

I just need to think about how I can setup the link between the two repo properly and without too much hassle.

**Comment by ImFlog** [comment]
It finally works @chriskn 🎉 
Can you make a sign if you get some time to test this new version on your real project ?
I have one week after the publication of a new release on the gradle plugin portal before it's boring to remove so Ideally I first get a sign from you and then I push the new version.
If Ok, I will delete the version and push this one without the SNAPSHOT suffix.
Thanks

**Comment by ImFlog** [comment]
Just published version `1.11.1`. 
Keep me posted 🙏

**Comment by chriskn** [comment]
Hi @ImFlog, since you hosted the dependency `avro-1.12.0-ImFlog` on GitHub (username and password required to retrieve the dependency), I can't build and publish a local release to test this.

**Comment by ImFlog** [comment]
Hi @chriskn, I've spent the night fighting with Maven and jitpack to finally be able to make the branch compile without the need for custom authentication on github registry.

If you add the repository `maven("https://jitpack.io")` in your buildScript you should be able to use the new version (still on a temporary branch on my avro fork repository).

Not a big fan of how I had to make it work as it will be hard to keep up with avro updates with it but I am first trying to assess if everything works well and fixing the current issues before thinking about how I could improve it.

Happy to get feedback, I would like to put this behind me before my holidays :)

**Comment by chriskn** [comment]
When testing this branch, I am getting the following exception

```
Caused by: com.github.imflog.avro.SchemaParseException: Undefined name: "<MyReferencedSchema>"
	at com.github.imflog.avro.Schema.parse(Schema.java:1825)
	at com.github.imflog.avro.Schema.parse(Schema.java:1915)
	at com.github.imflog.avro.Schema.parse(Schema.java:1872)
	at com.github.imflog.avro.Schema$Parser.parse(Schema.java:1610)
	at com.github.imflog.avro.Schema$Parser.parse(Schema.java:1596)
	at com.github.imflog.schema.registry.parser.AvroSchemaParser.resolveLocalReferences(AvroSchemaParser.kt:27)
	at com.github.imflog.schema.registry.parser.SchemaParser.parseSchemaFromFile(SchemaParser.kt:39)
	at com.github.imflog.schema.registry.tasks.compatibility.CompatibilityTaskAction.run(CompatibilityTaskAction.kt:29)
	at com.github.imflog.schema.registry.tasks.compatibility.CompatibilityTask.testCompatibility(CompatibilityTask.kt:44)
```

Looks like referenced schemas are not correctly resolved when used in an array

**Comment by chriskn** [comment]
Hi @ImFlog, I added to failing test demonstrating the issue. It's not limited to arrays

**Comment by ImFlog** [comment]
Sorry for the late response @chriskn, this is finally working. While refactoring I forgot to parse the local dependency in "reverse" as stated in the documentation.
Both of your tests work correctly now.
If you have no other failing case, I will clean up my custom avro version and update the PR accordingly.
After that I will be able to merge this and finally move on to other things 😅
